### PR TITLE
Adds support for USB HS and Mass Storage Device

### PR DESCRIPTION
**Pull request by bugobliterator**
* Tested working on HereProAP board

With OTG HS write to emmc hovers around 700KB/s and Read Speed around 3MB/s

### CODE REVIEW

**Comment by tridge** [comment]
looks good, just need to look at size impact

File: libraries/AP_HAL_ChibiOS/hwdef/HereProAP/scripts/hereproled.lua [hunk 1]
```diff
@@ -1,5 +1,4 @@
 -- This script is a test of led override
--- luacheck: only 0
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/HereProAP/scripts/hereproled.lua:2]: Should you be completely removing this?

File: libraries/AP_HAL_ChibiOS/hwdef/common/chibios_board.mk [hunk 2]
```diff
@@ -133,6 +133,7 @@ CSRC += $(HWDEF)/common/stubs.c \
 	   $(HWDEF)/common/usbcfg.c \
 	   $(HWDEF)/common/usbcfg_dualcdc.c \
 	   $(HWDEF)/common/usbcfg_common.c \
+      $(HWDEF)/common/usbcfg_cdc_msd.c \
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/chibios_board.mk:136]: Can you get the formatting correct? I cleaned this up with my last PR

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32h7_mcuconf.h [hunk 3]
```diff
@@ -607,12 +607,16 @@
...
 #define STM32_USB_USE_OTG1                  TRUE
+#endif
+#ifndef STM32_USB_USE_OTG2
 #define STM32_USB_USE_OTG2                  TRUE
+#endif
 #define STM32_USB_OTG1_IRQ_PRIORITY         14
 #define STM32_USB_OTG2_IRQ_PRIORITY         14
-#define STM32_USB_OTG1_RX_FIFO_SIZE         512
-#define STM32_USB_OTG2_RX_FIFO_SIZE         1024
+#define STM32_USB_OTG1_RX_FIFO_SIZE         2048
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32h7_mcuconf.h:618]: This is a significant change, think it should be dependent on using OTG_HS

File: libraries/AP_HAL_ChibiOS/hwdef/common/usbcfg.c [hunk 4]
```diff
@@ -401,14 +401,14 @@ const USBConfig usbcfg = {
  * Serial over USB driver configuration.
  */
 const SerialUSBConfig serusbcfg1 = {
-#if STM32_OTG2_IS_OTG1
-  &USBD2,
-#else
+#if STM32_USB_USE_OTG1 || STM32_OTG2_IS_OTG1
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/usbcfg.c:404]: This is wrong - you will have broken STM32H730

**Review by andyp1per — DISMISSED** [review]
Needs testing on SPRacingH7RF - a couple of things look wrong

File: libraries/AP_HAL_ChibiOS/hwdef/common/chibios_board.mk (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/chibios_board.mk]: there's still tabs in this file though, I have made changes to be same, but I don't think its still clean.

File: libraries/AP_HAL_ChibiOS/hwdef/HereProAP/scripts/hereproled.lua (hunk 1 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/HereProAP/scripts/hereproled.lua:2]: issues fixed

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32h7_mcuconf.h (hunk 3 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/stm32h7_mcuconf.h:618]: for stm32h7 chips this is not a huge cost, I don't think its needed to have ifdefs around here.

File: libraries/AP_HAL_ChibiOS/hwdef/common/usbcfg.c (hunk 4 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/usbcfg.c]: should be fixed, can you check now

File: libraries/AP_HAL_ChibiOS/sdcard.cpp [hunk 5]
```diff
@@ -47,6 +62,11 @@ static SPIConfig lowspeed;
 static SPIConfig highspeed;
 #endif
 
+#if HAL_HAVE_USB_CDC_MSD
+static uint8_t blkbuf[512];
```
[the remark this answers is not in the document]
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/sdcard.cpp:66]: Its used as a bounce buffer, its not used with DMA.

[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/sdcard.cpp:66]: ok, but still think these should be dynamically allocated like we do with the other bouncebuffers

**Comment by andyp1per** [comment]
Tested on SPRacingH7RF

**Comment by magicrub** [comment]
Is this H757 core clock running at 480MHz? If not, would it be possible to bump it up during transfers to see if that results in higher data throughput? I'm also thinking Networking could benefit from that scheme.. In theory the USB 2.0 HS should be able to push 480 Mbps (60 MB/s) but you stated you're seeing 3. So, somewhere there's a bottleneck. Is this reading from an SD card?

**Comment by magicrub** [comment]
@bugobliterator this isn't building any more

File: libraries/AP_HAL_ChibiOS/sdcard.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/sdcard.cpp:69]: at least do uint32_t and 512/4

**Review by andyp1per — APPROVED** [review]

**Pull request merged**